frontera/contrib/backends/cassandra/components.py

```diff
@@ -0,0 +1,359 @@
+        m = self._create_page(page)
+        m.error = error
+        self.cache[m.fingerprint] = m
+        query_page = self.session.prepare(
+            "UPDATE metadata SET error = ? WHERE fingerprint = ?")
+        self.session.execute(query_page, (error, page.meta['fingerprint']))
+        self.cass_count({"error": 1})
+
+    @retry_and_rollback
+    def page_crawled(self, response, links):
+        query_page = self.session.prepare(
+            "UPDATE metadata SET fetched_at = ?, headers = ?, method = ?, cookies = ?, status_code = ? "
+            "WHERE crawl= ? AND fingerprint = ?")
+        self.session.execute_async(query_page, (datetime.utcnow(), response.request.headers, response.request.method,
+                                                response.request.cookies, response.status_code, self.crawl_id,
+                                                response.meta['fingerprint']))
+
+        query = self.session.prepare(
+            "INSERT INTO metadata (crawl, fingerprint, created_at, method, url, depth) VALUES (?, ?, ?, ?, ?, ?)")
+        cql_items = []
+        for link in links:
+            if link.meta['fingerprint'] not in self.cache:
+                link.depth = self.cache[response.meta['fingerprint']].depth+1
+                l = self._create_page(link)
+                self.cache[link.meta['fingerprint']] = l
+                cql_i = (self.crawl_id, link.meta['fingerprint'], datetime.utcnow(), link.method, link.url, link.depth)
+                cql_items.append(cql_i)
+        execute_concurrent_with_args(self.session, query, cql_items, concurrency=400)
+        self.cass_count({"pages_crawled": 1, "links_found": len(cql_items)})
+
+    def _create_page(self, obj):
+        db_page = self.model()
+        db_page.fingerprint = obj.meta['fingerprint']
+        db_page.url = obj.url
+        db_page.created_at = datetime.utcnow()
+        db_page.meta = obj.meta
+        if hasattr(obj, 'depth'):
+            db_page.depth = obj.depth
+        else:
+            db_page.depth = 0
+
+        if isinstance(obj, Request):
+            db_page.headers = obj.headers
+            db_page.method = obj.method
+            db_page.cookies = obj.cookies
+        elif isinstance(obj, Response):
+            db_page.headers = obj.request.headers
+            db_page.method = obj.request.method
+            db_page.cookies = obj.request.cookies
+            db_page.status_code = obj.status_code
+        return db_page
+
+    @retry_and_rollback
+    def update_score(self, batch):
+        query = self.session.prepare("UPDATE metadata SET score = ? WHERE fingerprint = ?")
+        cql_items = []
+        for fprint, score, request, schedule in batch:
+            cql_i = (score, fprint)
+            cql_items.append(cql_i)
+        execute_concurrent_with_args(self.session, query, cql_items, concurrency=400)
+        self.cass_count({"scored_urls": len(cql_items)})
+
+    def cass_count(self, counts):
```

Comment:
I propose to make stats collection optional for the purpose of unification. This functionality is missing in other backends, and left to the user decide the implementation.  What do you think?